lib/core/test/unit/Cardano/Pool/DB/Mock.hs

```diff
@@ -0,0 +1,184 @@
+    , readPoolLifeCycleStatus
+        :: PoolId
+        -> m PoolLifeCycleStatus
+    , putPoolRegistration
+        :: CertificatePublicationTime
+        -> PoolRegistrationCertificate
+        -> m ()
+    , readPoolRegistration
+        :: PoolId
+        -> m (Maybe (CertificatePublicationTime, PoolRegistrationCertificate))
+    , putPoolRetirement
+        :: CertificatePublicationTime
+        -> PoolRetirementCertificate
+        -> m ()
+    , readPoolRetirement
+        :: PoolId
+        -> m (Maybe (CertificatePublicationTime, PoolRetirementCertificate))
+    , unfetchedPoolMetadataRefs
+        :: Int
+        -> m [(PoolId, StakePoolMetadataUrl, StakePoolMetadataHash)]
+    , putFetchAttempt
+        :: (StakePoolMetadataUrl, StakePoolMetadataHash)
+        -> m ()
+    , listRegisteredPools
+        :: m [PoolId]
+    , listRetiredPools
+        :: EpochNo
+        -> m [PoolRetirementCertificate]
+    , putPoolMetadata
+        :: StakePoolMetadataHash
+        -> StakePoolMetadata
+        -> m ()
+    , readPoolMetadata
+        :: m (Map StakePoolMetadataHash StakePoolMetadata)
+    , readSystemSeed
+        :: m StdGen
+    , rollbackTo
+        :: SlotNo
+        -> m ()
+    , removePools
+        :: [PoolId]
+        -> m ()
+    , cleanDB
+        :: m ()
+    }
+
+-- | An empty DB layer where none of the operations can be called.
+--
+dbLayer :: DBLayer m
+dbLayer = DBLayer
+    { cleanDB =
+        fail "cleanDB"
+    , listRegisteredPools =
+        fail "listRegisteredPools"
+    , listRetiredPools = \_ ->
+        fail "listRetiredPools"
+    , putFetchAttempt =
+        fail "putFetchAttempt"
+    , putPoolMetadata =
+        fail "putPoolMetadata"
+    , putPoolProduction =
+        fail "putPoolProduction"
+    , putPoolRegistration =
+        fail "putPoolRegistration"
+    , putPoolRetirement =
+        fail "putPoolRetirement"
+    , putStakeDistribution =
+        fail "putStakeDistribution"
+    , readPoolLifeCycleStatus =
+        fail "readPoolLifeCycleStatus"
+    , readPoolMetadata =
+        fail "readPoolMetadata"
+    , readPoolProduction =
+        fail "readPoolProduction"
+    , readPoolProductionCursor =
+        fail "readPoolProductionCursor"
+    , readPoolRegistration =
+        fail "readPoolRegistration"
+    , readPoolRetirement =
+        fail "readPoolRetirement"
+    , readStakeDistribution =
+        fail "readStakeDistribution"
+    , readSystemSeed =
+        fail "readSystemSeed"
+    , readTotalProduction =
+        fail "readTotalProduction"
+    , removePools =
+        fail "removePools"
+    , rollbackTo =
+        fail "rollbackTo"
+    , unfetchedPoolMetadataRefs =
+        fail "unfetchedPoolMetadataRefs"
+    }
+  where
+    fail op = error $ "mock DB layer: unexpected call to: " <> op
+
+-- | Converts a 'DBLayer' into a 'Real.DBLayer'.
+--
+toReal
+    :: (MonadFail m, MonadIO m)
+    => DBLayer m
+    -> Real.DBLayer m
+toReal DBLayer {..} = Real.DBLayer {..}
```

Comment:
Was going to say that defining a duplicate `DBLayer` seemed a bit overkill, but this `RecordWildCards` line is pretty neat!

Not 100% convinced, but keeping it up-to-date doesn't look _that_ annoying.

Reply:
> Was going to say that defining a duplicate DBLayer seemed a bit overkill

I was also reluctant to use this approach, but unfortunately, it's not possible to use record update syntax with record types that use existential quantification. See: https://gitlab.haskell.org/ghc/ghc/-/issues/16501

So the only option here is to define a simpler type where the record fields _can_ be updated, and then map values of that type back to the original type that uses existentials.

>  but this RecordWildCards line is pretty neat!

Thanks!

IMO, the advantage of using this method is that the compiler will tell us if we need to update the `Mock.DBLayer` type.